portal-kernel/src/com/liferay/portal/kernel/upgrade/UpgradeProcess.java

```diff
@@ -326,6 +336,62 @@ public boolean shouldDropIndex(Collection<String> columnNames) {
 
 	}
 
+	protected static <T, E1 extends Exception, E2 extends Exception> void
+			concurrentUpgrade(
+				UnsafeSupplier<T, E1> unsafeSupplier,
+				UnsafeConsumer<T, E2> unsafeConsumer)
+		throws E1, E2 {
+
+		Objects.requireNonNull(unsafeSupplier);
+		Objects.requireNonNull(unsafeConsumer);
+
+		ExecutorService executorService = Executors.newWorkStealingPool();
+
+		ThrowableCollector throwableCollector = new ThrowableCollector();
+
+		List<Future<Void>> futures = new ArrayList<>();
+
+		try {
+			long companyId = CompanyThreadLocal.getCompanyId();
+
+			T next = null;
+
+			while ((next = unsafeSupplier.get()) != null) {
+				T current = next;
+
+				Future<Void> future = executorService.submit(
+					() -> {
+						try (SafeCloseable safeCloseable =
+								CompanyThreadLocal.lock(companyId)) {
+
+							unsafeConsumer.accept(current);
+						}
+
+						return null;
+					});
+
+				futures.add(future);
+			}
+		}
+		finally {
+			executorService.shutdown();
+
+			for (Future<Void> future : futures) {
+				try {
+					future.get();
+				}
+				catch (ExecutionException executionException) {
+					throwableCollector.collect(executionException.getCause());
+				}
+				catch (Exception exception) {
+					throwableCollector.collect(exception);
+				}
+			}
+		}
+
+		throwableCollector.rethrow();
```

Comment:
The exception management has changed. In most of the multithreading upgrades we caught the exceptions in the callable, print an error, and return false. Then, we got the returned values from the future and if one of them was false we threw an exception. With this change, we won't throw an exception in this case.

Reply:
ThrowableCollector is thread safe, if it makes more sense to you we can move the throwableCollector.collect(exception) into the callable that we are submitting.

Reply:
I get it now, it's because the existing callables are all catching the exception. This is fixed in 7c77b73 and 837e6dc.